### PR TITLE
Add records support in navigate to

### PR DESCRIPTION
**Pull request by Youssef1313**
Before:

image

image

After:

image

image

### CODE REVIEW

File: src/Tools/ExternalAccess/FSharp/NavigateTo/FSharpNavigateToItemKind.cs [hunk 1]
```diff
@@ -11,6 +11,7 @@ internal static class FSharpNavigateToItemKind
         public static string Line => NavigateToItemKind.Line;
         public static string File = NavigateToItemKind.File;
         public static string Class => NavigateToItemKind.Class;
+        public static string Record => NavigateToItemKind.Record;
```
Comment (Youssef1313) [inline src/Tools/ExternalAccess/FSharp/NavigateTo/FSharpNavigateToItemKind.cs:14]: This class is used in dotnet/fsharp repository. I think we need to check with the other team whether they want this or not?

File: src/Features/Core/Portable/Navigation/NavigableItemFactory.DeclaredSymbolNavigableItem.cs [hunk 2]
```diff
@@ -45,6 +45,7 @@ private static Glyph GetPublicGlyph(DeclaredSymbolInfoKind kind)
                 switch (kind)
                 {
                     case DeclaredSymbolInfoKind.Class: return Glyph.ClassPublic;
+                    //case DeclaredSymbolInfoKind.Record: return Glyph.RecordPublic; // I'm not sure what a Glyph is? this I add "RecordPublic" to the enum?
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/Navigation/NavigableItemFactory.DeclaredSymbolNavigableItem.cs:48]:
```suggestion
                    case DeclaredSymbolInfoKind.Record: return Glyph.ClassPublic;
```

Comment (CyrusNajmabadi) [reply src/Features/Core/Portable/Navigation/NavigableItemFactory.DeclaredSymbolNavigableItem.cs:48]: yup.  Add RecordPublic to that enum

File: src/Tools/ExternalAccess/FSharp/NavigateTo/FSharpNavigateToItemKind.cs (hunk 1 above)
Comment (CyrusNajmabadi) [reply src/Tools/ExternalAccess/FSharp/NavigateTo/FSharpNavigateToItemKind.cs:14]: this is fine.

File: src/Workspaces/Core/Portable/FindSymbols/DeclaredSymbolInfo.cs [hunk 3]
```diff
@@ -30,6 +30,7 @@ internal enum DeclaredSymbolInfoKind : byte
         Module,
         Property,
         Struct,
+        Record,
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/FindSymbols/DeclaredSymbolInfo.cs:33]: i would keep these sorted.

File: src/Workspaces/Core/Portable/FindSymbols/FindReferences/DependentTypeFinder_ProjectIndex.cs [hunk 4]
```diff
@@ -65,6 +65,7 @@ private static async Task<ProjectIndex> CreateIndexAsync(Project project, Cancel
                     {
                         switch (info.Kind)
                         {
+                            // TODO: Not sure what is the correct action for records.
```
[the remark this answers is not in the document]
Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/FindSymbols/FindReferences/DependentTypeFinder_ProjectIndex.cs:68]: don't call it `typesThatMayDeriveFromSystemObject ` as that would be everything.  instead, call it classesOrRecordsThat...

File: src/EditorFeatures/CSharpTest/NavigateTo/NavigateToTests.cs [hunk 5]
```diff
@@ -55,6 +55,20 @@ public async Task FindClass(TestHost testHost)
...
+        [CombinatorialData]
+        public async Task FindRecord(TestHost testHost)
+        {
+            await TestAsync(
+testHost, @"record Goo
+{
+}", async w =>
+            {
+                var item = (await _aggregator.GetItemsAsync("Goo")).Single(x => x.Kind != "Method");
+                VerifyNavigateToResultItem(item, "Goo", "[|Goo|]", PatternMatchKind.Exact, NavigateToItemKind.Record, Glyph.ClassInternal);
```
Comment (Youssef1313) [inline src/EditorFeatures/CSharpTest/NavigateTo/NavigateToTests.cs:68]: The `NavigateToItemKind` used here doesn't have "Record" in it.
I think this one comes from a package reference "Microsoft.VisualStudio.Language.NavigateTo.Interfaces"

@CyrusNajmabadi What's the correct action here?
Is this package developed in this same repo or it's internal? Does it need to be updated first? Why we even need it if we have another NavigateToItemKind available?

**Comment by jcouv** [comment]
Thanks for fixing this. Could you add an entry to the [IDE test plan](https://github.com/dotnet/roslyn/blob/master/docs/wiki/Manual-Testing.md)? I wasn't aware this was a separate scenario.

File: src/EditorFeatures/CSharpTest/NavigateTo/NavigateToTests.cs (hunk 5 above)
Comment (CyrusNajmabadi) [reply src/EditorFeatures/CSharpTest/NavigateTo/NavigateToTests.cs:68]: We woudl need to map our type to something in the existing enum.  

> Is this package developed in this same repo or it's internal? 

It's part of VS, outside of Roslyn :)

> Does it need to be updated first?

Nope.  We can just map our concepts to theirs.  We can decide later if they need to expose the concept as well.

File: src/Features/Core/Portable/Navigation/NavigableItemFactory.DeclaredSymbolNavigableItem.cs (hunk 2 above)
Comment (Youssef1313) [reply src/Features/Core/Portable/Navigation/NavigableItemFactory.DeclaredSymbolNavigableItem.cs:48]: @CyrusNajmabadi What behavior would this introduce? I'm unsure of the other places that will need be updated.

Comment (CyrusNajmabadi) [reply src/Features/Core/Portable/Navigation/NavigableItemFactory.DeclaredSymbolNavigableItem.cs:48]: > In other words, do we really need to treat records differently here?

Yes, i woudl treat them differntly *within* roslyn.  at the boundaries with components out of our control, we may have to consider what to do.

Comment (Youssef1313) [reply src/Features/Core/Portable/Navigation/NavigableItemFactory.DeclaredSymbolNavigableItem.cs:48]: @CyrusNajmabadi Can this be addressed separately?

As I said, I'm not sure of the other places that needs to be updated. Also getting the Glyph from ISymbol is dependent on TypeKind enum currently. That should wait until https://github.com/dotnet/roslyn/issues/47713 is resolved.

Comment (CyrusNajmabadi) [reply src/Features/Core/Portable/Navigation/NavigableItemFactory.DeclaredSymbolNavigableItem.cs:48]: yes.  can you open an issue to track this.

File: src/EditorFeatures/CSharpTest/NavigateTo/NavigateToTests.cs (hunk 5 above)
Comment (Youssef1313) [reply src/EditorFeatures/CSharpTest/NavigateTo/NavigateToTests.cs:68]: @CyrusNajmabadi I'm not sure what you meant by this.

Comment (CyrusNajmabadi) [reply src/EditorFeatures/CSharpTest/NavigateTo/NavigateToTests.cs:68]: come to discord and we can discuss things there :)

File: src/EditorFeatures/Core.Wpf/NavigateTo/NavigateToItemProvider.Searcher.cs [hunk 6]
```diff
@@ -229,6 +229,48 @@ private void ReportMatchResult(Project project, INavigateToSearchResult result)
                 _callback.AddItem(navigateToItem);
             }
 
+            private static string GetKind(string kind)
```
Comment (Youssef1313) [inline src/EditorFeatures/Core.Wpf/NavigateTo/NavigateToItemProvider.Searcher.cs:232]: @CyrusNajmabadi This is per our discussion in discord.
I hope I did it the correct way.

File: src/Features/Core/Portable/Navigation/NavigableItemFactory.DeclaredSymbolNavigableItem.cs (hunk 2 above)
Comment (Youssef1313) [inline src/Features/Core/Portable/Navigation/NavigableItemFactory.DeclaredSymbolNavigableItem.cs:48]:
```suggestion
```

**Review by CyrusNajmabadi — APPROVED** [review]
Thanks!

**Pull request merged**